### PR TITLE
Dispatch ALTER DATABASE with options

### PR DESCRIPTION
**Pull request by soumyadeep2007**
Earlier, we used to dispatch CdbDispatchCommand() to dispatch ALTER DATABASEs with only the CONNECTION LIMIT parameter. The consequence of this was that other ALTER time options like ALLOW_CONNECTIONS and IS_TEMPLATE weren't, and the catalog changes involved would never be done on the segments.

Remedy that by dispatching the AlterDatabaseStmt node instead. Also, add support to serialize/deserialize this node to make the dispatch possible.

Changes to tests (connection_limits.sql):

* connection_limits was never in the schedule, probably because it didn't modify pg_hba.conf to allow connections from connlimit_test_user and there were hba errors. Remedy that by adding code to modify pg_hba.conf.

* Add coverage for ALLOW_CONNECTIONS and IS_TEMPLATE.

* Assert that catalog changes are reflected everywhere.

* Since we are expanding the scope of connection_limits.sql to cover for these cases, we rename it accordingly to dboptions.sql.

### CODE REVIEW

File: src/test/regress/sql/dboptions.sql [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+union
+select gp_segment_id, datconnlimit from gp_dist_random('pg_database') where datname like 'limit%db'
+order by gp_segment_id;
+
+-- clean up globals
+drop user connlimit_test_user;
+-- remove rule from pg_hba.conf for connlimit_test_user
+\! sed -i '$ d' $MASTER_DATA_DIRECTORY/pg_hba.conf
+select pg_reload_conf();
+drop database limitdb2;
```
Comment (ashwinstar) [inline src/test/regress/sql/dboptions.sql:86]: I feel we should leave these interesting database objects behind to validate dump/upgrade handle them fine. Thoughts?

**Review by ashwinstar — APPROVED** [review]

**Review by huansong — APPROVED** [review]
LGTM! Should backport it to 6X too?

File: src/test/regress/sql/dboptions.sql [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+union
+select gp_segment_id, datconnlimit from gp_dist_random('pg_database') where datname='limitdb'
+order by gp_segment_id;
+
+alter database limitdb with connection limit 0;
+
+-- create a regular user as superusers are exempt from limits
+create user connlimit_test_user;
+-- add superuser to pg_hba.conf
+\! echo "local    all         connlimit_test_user         trust" >> $MASTER_DATA_DIRECTORY/pg_hba.conf
```
Comment (l-wang) [inline src/test/regress/sql/dboptions.sql:23]: Nit: s/MASTER_DATA_DIRECTORY/COORDINATOR_DATA_DIRECTORY/

File: src/test/regress/sql/dboptions.sql [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+alter database "limit_evil_'""_db" connection limit 3;
+select -1 as gp_segment_id, datconnlimit from pg_database where datname like 'limit%db'
+union
+select gp_segment_id, datconnlimit from gp_dist_random('pg_database') where datname like 'limit%db'
+order by gp_segment_id;
+
+-- clean up globals
+drop user connlimit_test_user;
+-- remove rule from pg_hba.conf for connlimit_test_user
+\! sed -i '$ d' $MASTER_DATA_DIRECTORY/pg_hba.conf
```
Comment (l-wang) [inline src/test/regress/sql/dboptions.sql:84]: Nit: s/MASTER_DATA_DIRECTORY/COORDINATOR_DATA_DIRECTORY/

File: src/test/regress/sql/dboptions.sql (hunk 1 above)
Comment (l-wang) [reply src/test/regress/sql/dboptions.sql:86]: I like this idea!

**Review by l-wang — APPROVED** [review]

File: src/test/regress/sql/dboptions.sql (hunk 2 above)
Comment (soumyadeep2007) [reply src/test/regress/sql/dboptions.sql]: Done!

File: src/test/regress/sql/dboptions.sql (hunk 3 above)
Comment (soumyadeep2007) [reply src/test/regress/sql/dboptions.sql]: Done!

File: src/test/regress/sql/dboptions.sql (hunk 1 above)
Comment (soumyadeep2007) [reply src/test/regress/sql/dboptions.sql]: :+1: changed. Had to re-allow db connections to avoid a pg_upgrade --check failure.

**Comment by soumyadeep2007** [comment]
> LGTM! Should backport it to 6X too?

IS_TEMPLATE and ALLOW_CONNECTIONS was introduced 9.5+ and is thus N/A for 6X and 5X

**Pull request merged**